paramiko/auth_handler.py

```diff
@@ -229,6 +271,74 @@ def _parse_service_request(self, m):
         # dunno this one
         self._disconnect_service_not_available()
 
+    def _pkcs11_get_public_key(self):
+        public_key = ""
+        if self.pkcs11session is None:
+            if which("pkcs15-tool") is None:
```

Comment:
I think you should remove the `which()` function you added. Just catch the `subprocess.Popen()` exception for this case (an `OSError` with `.errno` equal to https://docs.python.org/2/library/errno.html#errno.ENOENT). Then you can re-throw your more specific `Exception("pkcs15-tool not found in PATH")` (I wouldn't include the extended instructions in the exception text).